### PR TITLE
Compatibility with ChainRules, Drop Julia 1.0 support

### PR DESCRIPTION
**Pull request by rick2047**
related to #396. As mentioned by @moesphere in that thread, we need to define a rrule for the Interolations type. I've defined a simple rrule which enables me to integrate with, for example, Zygote like this:

```julia
julia> using Interpolations

julia> y = sin.(1:10)
10-element Vector{Float64}:
  0.8414709848078965
  0.9092974268256817
  0.1411200080598672
 -0.7568024953079282
 -0.9589242746631385
 -0.27941549819892586
  0.6569865987187891
  0.9893582466233818
  0.4121184852417566
 -0.5440211108893698

julia> itp = interpolate(y,BSpline(Cubic(Reflect(OnCell()))))
10-element interpolate(OffsetArray(::Vector{Float64}, 0:11), BSpline(Cubic(Reflect(OnCell())))) with element type Float64:
  0.7125302276762112
  0.769963450028415
  0.1194958272928704
 -0.6408357079724973
 -0.8119858486932346
 -0.23659994479000876
  0.556314857216602
  0.8377563450756788
  0.3489685127835064
 -0.29399432638595374

julia> Zygote.gradient(itp, 2)
([-0.35017548837401463],)

julia> Interpolations.gradient(itp, 2)
1-element StaticArrays.SVector{1, Float64} with indices SOneTo(1):
 -0.35017548837401463
```

This should allow us to integrate with Flux.

### CODE REVIEW

**Comment by mkitti** [comment]
Awesome. What's the best way to test this?

**Comment by rick2047** [comment]
Actually I was going to ask the community the same question. I can of course write a naive test which takes gradient over a known function and we compare values. But that isn't the best approach I think. What we really want to do is check integration with chain rules.

The discussion started with trying to implement upsample methods. Maybe we can do something with that?

**Comment by rick2047** [comment]
I found [ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl) I guess that's the best approach.

**Comment by rick2047** [comment]
I tried using ChainRulesTestUtils but it seems like it doesn't work for Interpolation objects. Adding code here in case I am doing something obviously wrong.
```julia
julia> using ChainRulesTestUtils

julia> x = collect(1:10)
julia> y = sin.(x)

julia> itp = interpolate(y,BSpline(Linear()))


julia> test_rrule(itp, 1, 1)
test_rrule: [0.8414709848078965, 0.9092974268256817, 0.1411200080598672, -0.7568024953079282, -0.9589242746631385, -0.27941549819892586, 0.6569865987187891, 0.9893582466233818, 0.4121184852417566, -0.5440211108893698] at (1, 1): Error During Test at C:\Users\paresh\.julia\packages\ChainRulesTestUtils\NCO8i\src\testers.jl:168
  Got exception outside of a @test
  ArgumentError: test_rrule cannot be used on closures/functors (such as [0.8414709848078965, 0.9092974268256817, 0.1411200080598672, -0.7568024953079282, -0.9589242746631385, -0.27941549819892586, 0.6569865987187891, 0.9893582466233818, 0.4121184852417566, -0.5440211108893698])
  Stacktrace:
    [1] _ensure_not_running_on_functor(f::Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear}, Tuple{Base.OneTo{Int64}}}, name::String)
      @ ChainRulesTestUtils ~\.julia\packages\ChainRulesTestUtils\NCO8i\src\testers.jl:238
    [2] macro expansion
      @ ~\.julia\packages\ChainRulesTestUtils\NCO8i\src\testers.jl:169 [inlined]
    [3] macro expansion
      @ C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\Test\src\Test.jl:1151 [inlined]
    [4] test_rrule(::Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear}, Tuple{Base.OneTo{Int64}}}, ::Int64, ::Vararg{Int64, N} 
where N; output_tangent::ChainRulesTestUtils.Auto, fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, check_inferred::Bool, fkwargs::NamedTuple{(), Tuple{}}, rtol::Float64, atol::Float64, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
      @ ChainRulesTestUtils ~\.julia\packages\ChainRulesTestUtils\NCO8i\src\testers.jl:169
    [5] test_rrule(::Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear}, Tuple{Base.OneTo{Int64}}}, ::Int64, ::Vararg{Int64, N} 
where N)
      @ ChainRulesTestUtils ~\.julia\packages\ChainRulesTestUtils\NCO8i\src\testers.jl:166
    [6] top-level scope
      @ REPL[32]:1
    [7] eval
      @ .\boot.jl:360 [inlined]
    [8] eval
      @ .\Base.jl:39 [inlined]
    [9] repleval(m::Module, code::Expr, #unused#::String)
      @ VSCodeServer ~\.vscode\extensions\julialang.language-julia-1.1.38\scripts\packages\VSCodeServer\src\repl.jl:124
   [10] (::VSCodeServer.var"#47#49"{Module, Expr, REPL.LineEditREPL, REPL.LineEdit.Prompt})()
      @ VSCodeServer ~\.vscode\extensions\julialang.language-julia-1.1.38\scripts\packages\VSCodeServer\src\repl.jl:99
   [11] with_logstate(f::Function, logstate::Any)
      @ Base.CoreLogging .\logging.jl:491
   [12] with_logger
      @ .\logging.jl:603 [inlined]
   [13] (::VSCodeServer.var"#46#48"{Module, Expr, REPL.LineEditREPL, REPL.LineEdit.Prompt})()
      @ VSCodeServer ~\.vscode\extensions\julialang.language-julia-1.1.38\scripts\packages\VSCodeServer\src\repl.jl:100
   [14] #invokelatest#2
      @ .\essentials.jl:708 [inlined]
   [15] invokelatest(::Any)
      @ Base .\essentials.jl:706
   [16] macro expansion
      @ ~\.vscode\extensions\julialang.language-julia-1.1.38\scripts\packages\VSCodeServer\src\eval.jl:34 [inlined]
   [17] (::VSCodeServer.var"#60#61")()
      @ VSCodeServer .\task.jl:406
Test Summary:
                                                                      | Error  Total
test_rrule: [0.8414709848078965, 0.9092974268256817, 0.1411200080598672, -0.7568024953079282, -0.9589242746631385, -0.27941549819892586, 0.6569865987187891, 0.9893582466233818, 0.4121184852417566, -0.5440211108893698] at (1, 1) |     1      1
ERROR: Some tests did not pass: 0 passed, 0 failed, 1 errored, 0 broken.
```

So this seems like a dud, any other ideas about testing are welcome

File: src/chainrules/chainrules.jl [hunk 1]
```diff
@@ -0,0 +1,8 @@
+export rrule
+function ChainRulesCore.rrule(itp::AbstractInterpolation, x...)
+    y = itp(x...)
+    function back(Δ)
+        (NO_FIELDS, Δ * gradient(itp, x...))
```
Comment (DhairyaLGandhi) [inline src/chainrules/chainrules.jl:5]: Would you not want to update any of the fields of the interpolation too?

**Comment by DhairyaLGandhi** [comment]
Yay! This looks like a really nice step. Maybe qualify the call to gradient as `Interpolations.gradient` for readability if nothing else?

**Comment by DhairyaLGandhi** [comment]
I would directly test that zygote is producing the correct and expected value

File: src/chainrules/chainrules.jl (hunk 1 above)
Comment (rick2047) [reply src/chainrules/chainrules.jl:5]: I am not sure about that. I am confident that the second argument needs to be the gradient, but I am not sure what the first argument is even supposed to be. I have hacked this in to see if it works but I am not confident this is right. What would you recommend? Maybe you can clarify a bit more on what the first and second argument are supposed to be

Comment (rick2047) [reply src/chainrules/chainrules.jl:5]: I will change the call to Interpolations.gradient, I made the choice at random anyways.

**Comment by rick2047** [comment]
@DhairyaLGandhi thanks for your attention here. I have added a naive test, if this works we can think about adding more comprehensive testing.

**Comment by rick2047** [comment]
I dug around a bit more on ChainRulesTestUtils, they already have an issue open for the functor issue https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/117#issue-801431029

**Comment by rick2047** [comment]
I have no idea why these tests are failing. Did I do dependencies wrong?

**Comment by mkitti** [comment]
```julia
Interpolations: Error During Test at D:\a\Interpolations.jl\Interpolations.jl\test\chainrules.jl:10
  Test threw exception
  Expression: (Zygote.gradient(itp, 1))[1] == Interpolations.gradient(itp, 1)
  ArgumentError: unable to check bounds for indices of type Interpolations.WeightedAdjIndex{2,Int64}
```

**Comment by rick2047** [comment]
@mkitti  I saw that, that is the original error we started to solve. I don't understand how it works on 3 tests and doesn't work on 3 tests. 

also, itworksonmymachine.jpg :(

**Comment by mkitti** [comment]
```julia
julia> Zygote.gradient(itp,1)
ERROR: ArgumentError: unable to check bounds for indices of type Interpolations.WeightedAdjIndex{2, Int64}
Stacktrace:
  [1] checkindex(#unused#::Type{Bool}, inds::Base.OneTo{Int64}, i::Interpolations.WeightedAdjIndex{2, Int64})
    @ Base .\abstractarray.jl:671
  [2] checkbounds
    @ .\abstractarray.jl:601 [inlined]
  [3] checkbounds
    @ .\abstractarray.jl:616 [inlined]
  [4] view
    @ .\subarray.jl:177 [inlined]
  [5] (::Zygote.var"#393#395"{Vector{Float64}, Tuple{Interpolations.WeightedAdjIndex{2, Int64}}})(dy::Float64)
    @ Zygote ~\.julia\packages\Zygote\RxTZu\src\lib\array.jl:41
  [6] (::Zygote.var"#2220#back#389"{Zygote.var"#393#395"{Vector{Float64}, Tuple{Interpolations.WeightedAdjIndex{2, Int64}}}})(Δ::Float64)
    @ Zygote ~\.julia\packages\ZygoteRules\OjfTt\src\adjoint.jl:59
  [7] (::Zygote.var"#178#179"{Tuple{Tuple{Nothing}, Tuple{Nothing}}, Zygote.var"#2220#back#389"{Zygote.var"#393#395"{Vector{Float64}, Tuple{Interpolations.WeightedAdjIndex{2, Int64}}}}})(Δ::Float64)
    @ Zygote ~\.julia\packages\Zygote\RxTZu\src\lib\lib.jl:194
  [8] (::Zygote.var"#1686#back#180"{Zygote.var"#178#179"{Tuple{Tuple{Nothing}, Tuple{Nothing}}, Zygote.var"#2220#back#389"{Zygote.var"#393#395"{Vector{Float64}, Tuple{Interpolations.WeightedAdjIndex{2, Int64}}}}}})(Δ::Float64)
    @ Zygote ~\.julia\packages\ZygoteRules\OjfTt\src\adjoint.jl:59
  [9] Pullback
    @ ~\.julia\packages\Interpolations\qHlUr\src\b-splines\indexing.jl:8 [inlined]
 [10] (::Zygote.var"#41#42"{typeof(∂(λ))})(Δ::Float64)
    @ Zygote ~\.julia\packages\Zygote\RxTZu\src\compiler\interface.jl:41
 [11] gradient(f::Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear}, Tuple{Base.OneTo{Int64}}}, args::Int64)
    @ Zygote ~\.julia\packages\Zygote\RxTZu\src\compiler\interface.jl:59
 [12] top-level scope
    @ REPL[70]:1
```

**Comment by mkitti** [comment]
Ah I see. The tests are failing on Julia 1.0.5. They succeed on Julia 1.6.

**Comment by mkitti** [comment]
Where the tests are failing:

```
 Installed Zygote ─────────────── v0.4.20
 Installed ZygoteRules ────────── v0.2.1
 ```

**Comment by rick2047** [comment]
I added another test for 2D interpolation case. Now I am a bit stuck about how should we proceed. @DhairyaLGandhi  is still scrutinizing the actual rule, but apart from that it seems like integration with Zygote is working. Ideas on how to proceed are welcome, I was thinking maybe we can write what is asked for in this issue https://github.com/JuliaImages/ImageTransformations.jl/issues/113?

**Comment by DhairyaLGandhi** [comment]
What else is needed right now? I am thinking it might be good to merge this and move from there. What do you think @mkitti ?

**Comment by mkitti** [comment]
The new code is in a mergeable state. Some housekeeping is required.

If we're going to break compat with Julia 1.0, this probably should be v0.14.0. However, it's only clear to me that we need Julia 1.3 for testing.

Do we have any bounds on compat with ChainRulesCore.jl or should we just declare compat against the latest version?

**Comment by DhairyaLGandhi** [comment]
The current minor version of ChainRulesCore should be fine. You can also make a patch release, it's allowed within semver.

**Comment by rick2047** [comment]
If this is mergeable thats do it. this doesn't break anything and adds new functionality. The only apprehension I had was if our rule, which basically is `Zygote.gradient(itp,x)[1] == Interpolations.gradient(itp,x)` the right definition. Should I document this somewhere?

**Comment by mkitti** [comment]
Well the tests do break Julia 1.0 compatibility, which is fine. It just needs to be managed. I'll probably run a v0.13 branch that is compatible with Julia 1.0 for a while longer in parallel with a v0.14 branch that requires Julia 1.3.

Documentation would be great. Could you add a `docs/src/chainrules.md` to describe this?

File: src/chainrules/chainrules.jl [hunk 2]
```diff
@@ -0,0 +1,12 @@
+export rrule
+"""
+    ChainRulesCore.rrule(itp::AbstractInterpolation, x...)
+ChainRulesCore.jl `rrule` for integration with automatic differentiation libraries.
+"""
+function ChainRulesCore.rrule(itp::AbstractInterpolation, x...)
+    y = itp(x...)
+    function pullback(Δy)
+        (NO_FIELDS, Δy * Interpolations.gradient(itp, x...))
```
Comment (DhairyaLGandhi) [inline src/chainrules/chainrules.jl:9]: One quick thought (sorry this should have come earlier) `NO_FIELDS` -> `DoesNotExist()`. @rick2047 would you be up for a tiny follow up?

Comment (rick2047) [reply src/chainrules/chainrules.jl:9]: sure, I was bothered by this too. but this is already merged, can I still push to this branch?

Comment (mkitti) [reply src/chainrules/chainrules.jl:9]: You can use the same branch. It's your branch. Just send a new PR when you are ready.

**Comment by mkitti** [comment]
We seem to be having CI issues with ChainRulesCore.jl 0.10. @rick2047 @DhairyaLGandhi , do you have any ideas where the issue may be?

**Pull request merged**